src/middleware.js

```diff
@@ -2,11 +2,35 @@ import path from 'path';
 
 import mime from 'mime-types';
 
+import getETag from 'etag';
+
 import getFilenameFromUrl from './utils/getFilenameFromUrl';
 import handleRangeHeaders from './utils/handleRangeHeaders';
 import ready from './utils/ready';
 
+const { hasOwnProperty } = Object.prototype;
+
 export default function wrapper(context) {
+  let etagRegistry;
+
+  if (context.options.etags) {
+    etagRegistry = new Map();
+    context.compiler.hooks.done.tap('webpack-dev-middleware', (stats) => {
+      etagRegistry.clear();
+
+      try {
+        const { assets } = stats.compilation;
+        for (const assetId in assets) {
+          if (hasOwnProperty.call(assets, assetId)) {
+            const { existsAt: fsPath } = assets[assetId];
+            const etag = getETag(assets[assetId].source());
```

Comment:
Seems this is no longer possible in Webpack 5, due to error:

> <e> [webpack-dev-middleware] Error: Content and Map of this Source is not available (only size() is supported)
> <e>     at SizeOnlySource._error (/node_modules/webpack-sources/lib/SizeOnlySource.js:16:10)
> <e>     at SizeOnlySource.source (/node_modules/webpack-sources/lib/SizeOnlySource.js:26:14)

@alexander-akait Do you have a recommended work around for this?

Reply:
I ended up leveraging `chunkHashs`, which is actually more convenient because I can just use the hash as the etag instead of generating one from the file source:

```js
const { chunkHashs: assetHashes } = stats.compilation.records;
for (const assetId in assetHashes) {
	if (hasOwnProp(assetHashes, assetId)) {
		etagRegistry.set(assetId, JSON.stringify(assetHashes[assetId]));
	}
}
```

If there's an alternative you recommend, LMK.

Reply:
/cc @sokra What do you think?